src/rgw/driver/rados/rgw_zone.h

```diff
@@ -525,47 +525,56 @@ WRITE_CLASS_ENCODER(RGWPeriodConfig)
 class RGWRealm;
 class RGWPeriod;
 
-class RGWRealm : public RGWSystemMetaObj
+class RGWRealm
 {
 public:
+  std::string id;
+  std::string name;
+
+  CephContext *cct{nullptr};
+
   std::string current_period;
   epoch_t epoch{0}; //< realm epoch, incremented for each new period
 
-  int create_control(const DoutPrefixProvider *dpp, bool exclusive, optional_yield y);
-  int delete_control(const DoutPrefixProvider *dpp, optional_yield y);
 public:
   RGWRealm() {}
-  RGWRealm(const std::string& _id, const std::string& _name = "") : RGWSystemMetaObj(_id, _name) {}
-  RGWRealm(CephContext *_cct, RGWSI_SysObj *_sysobj_svc): RGWSystemMetaObj(_cct, _sysobj_svc) {}
-  RGWRealm(const std::string& _name, CephContext *_cct, RGWSI_SysObj *_sysobj_svc): RGWSystemMetaObj(_name, _cct, _sysobj_svc){}
-  virtual ~RGWRealm() override;
+  RGWRealm(const std::string& _id, const std::string& _name = "") : id(_id), name(_name) {}
+  RGWRealm(CephContext *_cct): cct(_cct) {}
+  RGWRealm(const std::string& _name, CephContext *_cct, RGWSI_SysObj *_sysobj_svc): name(_name), cct(_cct){}
 
-  void encode(bufferlist& bl) const override {
+  const std::string& get_name() const { return name; }
+  const std::string& get_id() const { return id; }
+
+  void set_name(const std::string& _name) { name = _name;}
+  void set_id(const std::string& _id) { id = _id;}
+  void clear_id() { id.clear(); }
+
+  virtual ~RGWRealm();
+
+  void encode(bufferlist& bl) const {
     ENCODE_START(1, 1, bl);
-    RGWSystemMetaObj::encode(bl);
+    encode(id, bl);
+    encode(name, bl);
```

Comment:
you've correctly copied these from `RGWSystemMetaObj::encode()`, but that has an extra `ENCODE_START(1, 1, bl);` and `ENCODE_FINISH(bl);` wrapping them. so i think we need something like this instead:
```diff
   void encode(bufferlist& bl) const {
     ENCODE_START(1, 1, bl);
-    RGWSystemMetaObj::encode(bl);
+    {
+      // these used to be wrapped by RGWSystemMetaObj::encode(),
+      // so the extra ENCODE_START/ENCODE_FINISH are preserved
+      ENCODE_START(1, 1, bl);
+      encode(id, bl);
+      encode(name, bl);
+      ENCODE_FINISH(bl);
+    }
```
and similar for decode() below and DECODE_START/FINISH below